libraries/lib-snapping/SnapUtils.h

```diff
@@ -0,0 +1,131 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  SnapUtils.h
+
+  Dmitry Vedenko
+
+**********************************************************************/
+#pragma once
+
+#include "Prefs.h"
+#include "Registry.h"
+
+#include <functional>
+#include <string_view>
+
+enum class SnapMode
+{
+   SNAP_OFF,
+   SNAP_NEAREST,
+   SNAP_PRIOR
+};
+
+extern SNAPPING_API StringSetting SnapToSetting;
+extern SNAPPING_API EnumSetting<SnapMode> SnapModeSetting;
+
+SNAPPING_API SnapMode ReadSnapMode();
+SNAPPING_API Identifier ReadSnapTo();
+
+struct SnapRegistryGroup;
+struct SnapRegistryItem;
+
+struct SNAPPING_API SnapRegistryVisitor /* not final */
+{
+   virtual void BeginGroup(const SnapRegistryGroup& item) = 0;
+   virtual void EndGroup(const SnapRegistryGroup& item) = 0;
+   virtual void Visit(const SnapRegistryItem& item) = 0;
+};
+
+struct SNAPPING_API SnapConfig final
+{
+   double rate {};
+   double tempo {};
+   std::pair<int, int> timeSignature {};
+};
+
+struct SNAPPING_API SnapResult final
+{
+   //! Snapped time.
+   /*!
+      In case the snapping has failed for any reason, this will match the input.
+      Snapping can fail if the function matching the identifier was not found
+      or if the function decides the input is incorrect.
+      */
```

Comment:
Thanks for the explanation.

If the boolean means nothing more or less than a success indicator -- then I might instead just use `using SnapResult = std::optional<double>;`